dev/cuda/rmsnorm_backward.cu

```diff
@@ -0,0 +1,235 @@
+/*
+Kernels for layernorm backward pass.
+
+Compile example:
+nvcc -O3 --use_fast_math -lcublas -lcublasLt rmsnorm_backward.cu -o rmsnorm_backward
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <cuda_runtime.h>
+#include <assert.h>
+#include <cooperative_groups.h>
+#include <cooperative_groups/reduce.h>
+
+#include "common.h"
+
+// ----------------------------------------------------------------------------
+// CPU code reference
+
+void rmsnorm_forward_cpu(float* out, float* rstd,
+                         const float* inp, const float* weight,
+                         int B, int T, int C) {
+    float eps = 1e-5f;
+    for (int b = 0; b < B; b++) {
+        for (int t = 0; t < T; t++) {
+            // seek to the input position inp[b,t,:]
+            const float* x = inp + b * T * C + t * C;
+            // calculate the variance
+            float v = 0.0f;
+            for (int i = 0; i < C; i++) {
+                v += x[i] * x[i];
+            }
+            v = v/C;
+            // calculate the rstd (reciprocal standard deviation)
+            float s = 1.0f / sqrtf(v + eps);
+            // seek to the output position in out[b,t,:]
+            float* out_bt = out + b * T * C + t * C;
+            for (int i = 0; i < C; i++) {
+                float o = s * x[i] * weight[i];
+                out_bt[i] = o; // write
+            }
+            // cache the rstd for the backward pass later
+            rstd[b * T + t] = s;
+        }
+    }
+}
+
+void rmsnorm_backward_cpu(float* dinp, float* dweight,
+                          const float* dout, const float* inp, const float* weight, const float* rstd,
+                          int B, int T, int C) {
+    for (int b = 0; b < B; b++) {
+        for (int t = 0; t < T; t++) {
+            const float* dout_bt = dout + b * T * C + t * C;
+            const float* inp_bt = inp + b * T * C + t * C;
+            float* dinp_bt = dinp + b * T * C + t * C;
+            const float rstd_bt = rstd[b * T + t];
+
+            float o = 0.0f;
+            for (int i = 0; i < C; i++) {
+                o += weight[i] * dout_bt[i] * inp_bt[i];
```

Comment:
Thank you for this! As a result of this PR I have found a bug in my own RMSNorm impl. where I forgot to multiply by the weight in the backward pass. Spent days chasing it.

PS: I've ran this rmsnorm_backward_cpu through my test cases that verify against torch and all is passing.

Thanks again!

Reply:
void rms_backward(
            float* _In, float* _δ_Out, float* _δ_In,
            float* _δ_Gamma, float* _Gamma,
            int N, float EPSILON) {
            float rms = 0;
            for (int i = 0; i < N; i++) {
                rms += _In[i] * _In[i];
            }
            rms = std.sqrtf(rms / N + EPSILON);
            float rrms = 1.0f / rms;
            float rnrms = 1.0f / (N * rms * rms);
            float δ = 0;
            for (int i = 0; i < N; i++) {
                δ += _δ_Out[i] * _In[i] * rrms * **_Gamma[i]**; // This is what I missed!!
            }
            for (int i = 0; i < N; i++) {
                _δ_In[i] += _δ_Out[i] * _Gamma[i] * rrms - δ * _In[i] * rnrms;
                _δ_Gamma[i] += _δ_Out[i] * _In[i] * rrms;
            }
        }

Reply:
great to hear that this passes an independent set of tests. I did some testing, but wasn't yet fully sure that my implementation was correct.